diff -ruN a/blocks.h b/blocks.h
--- a/blocks.h	2020-04-01 11:56:48.542447768 +0300
+++ b/blocks.h	2020-04-01 12:21:56.343509591 +0300
@@ -1,20 +1,11 @@
-//Modify this file to change what commands output to your statusbar, and recompile using the make command.
-static const Block blocks[] = {
-	/*Icon*/	/*Command*/		/*Update Interval*/	/*Update Signal*/
-	{"", "cat ~/.pacupdate | sed /📦0/d",					0,		9},
-	
-	{"🧠", "free -h | awk '/^Mem/ { print $3\"/\"$2 }' | sed s/i//g",	30,		0},
-
-	{"", "~/bin/statusbar/volume",						0,		10},
-
-	{"☀", "xbacklight | sed 's/\\..*//'",					0,		11},
-	
-	{"", "~/bin/statusbar/battery",						5,		0},
-
-	{"🌡", "sensors | awk '/^temp1:/{print $2}'",				5,		0},
-
-	{"", "~/bin/statusbar/clock",						5,		0},
-};
-
-//sets delimeter between status commands. NULL character ('\0') means no delimeter.
-static char delim = '|';
+//Modify this file to change what commands output to your statusbar, and recompile using the make command.
+static const Block blocks[] = {
+	/*Icon*/	/*Command*/		/*Update Interval*/	/*Update Signal*/
+	{"",	"~/.local/bin/dwmblocks/newpackages",	0,	8},
+	{"",    "~/.local/bin/dwmblocks/volume",   0,  10},
+	{"",	"~/.local/bin/dwmblocks/battery | tr \'\n\' \' \'",	5,	0},
+	{"",	"date '+%H:%M%p'",	60,	0},
+};
+
+//sets delimeter between status commands. NULL character ('\0') means no delimeter.
+static char delim = '|';
