diff -rupN orig/contrib/dunst_xr_theme_changer.sh patched/contrib/dunst_xr_theme_changer.sh
--- orig/contrib/dunst_xr_theme_changer.sh	2021-04-22 08:32:36.373057664 +0300
+++ patched/contrib/dunst_xr_theme_changer.sh	2021-04-22 15:55:16.785146699 +0300
@@ -106,7 +106,7 @@ xrdb_get () {
 # Attributes dictionary. Add or remove attributes (see header for more info)
 #
 declare -A theme_attr_dict=(
-    ["global-font"]="$(xrdb_get 'font' 'Monospace') $(xrdb_get '*.font_size:' '11')"
+    ["global-font"]="$(xrdb_get 'font' 'Monospace') $(xrdb_get 'font_size' '11')"
     ["global-frame_width"]="$(xrdb_get 'border_width' '1')"
     ["global-frame_color"]="\"$(xrdb_get 'color8' '#65737e')\""
 
@@ -158,6 +158,19 @@ if ! [ -f "$user_conf" ]; then
 
         if [ -f "$example_conf_dir/dunstrc" ]; then
             example_conf="$example_conf_dir/dunstrc"
+        else
+            printf 'Could not find the example config file in "%s".
+               \nPlease, change $example_conf variable in the script.' \
+                   "$example_conf_dir"
+            exit 1
+        fi
+
+    elif [ -d "/etc/dunst" ]; then
+        # Void Linux default dir and example file
+        example_conf_dir="/etc/dunst"
+
+        if [ -f "$example_conf_dir/dunstrc" ]; then
+            example_conf="$example_conf_dir/dunstrc"
         else
             printf 'Could not find the example config file in "%s".
                \nPlease, change $example_conf variable in the script.' \
